### PR TITLE
Introduce ruff

### PR DESCRIPTION
**Pull request by sujuka99**
Replaces `isort`, `pyupgrade`, `flake8` with `ruff`

### CODE REVIEW

**Comment by sujuka99** [comment]
I would also like to add at some point [ANN](https://docs.astral.sh/ruff/rules/#flake8-annotations-ann) to ensure that we annotate our code consistently, but currently I don't have it configured properly locally and it's just annoying. I would have to play around with it for a bit before I can (maybe) reach an acceptable state where it is useful, but not inconvenient.

File: tests/component_handlers/schema_handler/test_schema_handler.py [hunk 1]
```diff
@@ -96,79 +96,91 @@ def test_should_lazy_load_schema_provider(find_class_mock: MagicMock):
...
+            id="components_module = None",
+        ),
+        pytest.param(
+            "",
+            id="components_module = ''",
+        ),
+    ],
+)
+def test_should_raise_value_error_when_schema_provider_is_called_and_components_module_is_empty(
+    components_module,
```
Comment (disrupted) [inline tests/component_handlers/schema_handler/test_schema_handler.py:149]: is there a rule to mark untyped parameters as warning?

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -35,11 +35,8 @@ jobs:
...
-      - name: Lint (flake8)
-        run: poetry run pre-commit run flake8 --all-files --show-diff-on-failure
-
-      - name: Order of imports (isort)
-        run: poetry run pre-commit run isort --all-files --show-diff-on-failure
+      - name: Lint (ruff)
+        run: poetry run pre-commit run ruff --all-files --show-diff-on-failure
 
       - name: Formatting (black)
         run: poetry run pre-commit run black --all-files --show-diff-on-failure
```
Comment (disrupted) [inline .github/workflows/ci.yaml:42]: what's your opinion on creating a separate branch to test `ruff format` so that we can see the diff? if it doesn't change between alpha and beta I'd say we can switch to it.

File: kpops/cli/main.py [hunk 3]
```diff
@@ -120,12 +122,17 @@ def setup_pipeline(
 
     handlers = setup_handlers(components_module, pipeline_config)
     return Pipeline.load_from_yaml(
-        pipeline_base_dir, pipeline_path, registry, pipeline_config, handlers
+        pipeline_base_dir,
+        pipeline_path,
+        registry,
+        pipeline_config,
+        handlers,
```
Comment (disrupted) [inline kpops/cli/main.py:129]: is this a new formatting rule?

File: kpops/components/base_components/base_defaults_component.py [hunk 4]
```diff
@@ -165,20 +170,23 @@ def defaults_from_yaml(path: Path, key: str) -> dict:
     """
     content = load_yaml_file(path, substitution=ENV)
     if not isinstance(content, dict):
-        raise TypeError(
+        msg = (
             "Default files should be structured as map ([app type] -> [default config]"
         )
+        raise TypeError(
+            msg,
```
Comment (disrupted) [inline kpops/components/base_components/base_defaults_component.py:177]: tbh not a fan of the forced trailing comma :( this could easily be a one liner

File: tests/component_handlers/schema_handler/test_schema_handler.py (hunk 1 above)
Comment (sujuka99) [reply tests/component_handlers/schema_handler/test_schema_handler.py:149]: `ANN`

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (sujuka99) [reply .github/workflows/ci.yaml:42]: Sounds good

File: kpops/components/base_components/base_defaults_component.py (hunk 4 above)
Comment (sujuka99) [reply kpops/components/base_components/base_defaults_component.py:177]: This probably happened because it used to be indented like that before. If I make it a one-liner it will not be flagged

File: kpops/cli/main.py (hunk 3 above)
Comment (sujuka99) [reply kpops/cli/main.py:129]: Nop, maybe something enforced line length :thinking:

File: tests/component_handlers/schema_handler/test_schema_handler.py (hunk 1 above)
Comment (disrupted) [reply tests/component_handlers/schema_handler/test_schema_handler.py:149]: ah ok, sorry I just saw your comment about it

Comment (disrupted) [reply tests/component_handlers/schema_handler/test_schema_handler.py:149]: otherwise I can be the one to enforce this rule manually 🤣

File: kpops/cli/main.py (hunk 3 above)
Comment (disrupted) [reply kpops/cli/main.py:129]: imo for function definitions with a lot of parameters this formatting is good, but function calls take up a lot of space this way

File: kpops/components/base_components/base_defaults_component.py (hunk 4 above)
Comment (sujuka99) [reply kpops/components/base_components/base_defaults_component.py]: Tested it now, `black` did it, because of the forced comma

Comment (sujuka99) [reply kpops/components/base_components/base_defaults_component.py]: This will be flagged and formatted by both `black` and `ruff format`
```
    def apply_component(
        self, component_class: type[PipelineComponent], component_data: dict,
    ) -> None:
```

This won't:
```
    def apply_component(
        self, component_class: type[PipelineComponent], component_data: dict
    ) -> None:
```

Comment (disrupted) [reply kpops/components/base_components/base_defaults_component.py]: do we have to force the trailing comma though?

File: tests/component_handlers/schema_handler/test_schema_handler.py (hunk 1 above)
Comment (disrupted) [reply tests/component_handlers/schema_handler/test_schema_handler.py:149]: alternatively, we could configure Pyright to warn on symbols where it needs a type annotation, i.e. in all cases where it can't be inferred.

File: kpops/components/base_components/base_defaults_component.py (hunk 4 above)
Comment (sujuka99) [reply kpops/components/base_components/base_defaults_component.py]: How can I revert just the [COM](https://github.com/bakdata/kpops/pull/363/commits/301284eb02dbb707fefa5ab50476b06ebee757ff) commit?

Comment (disrupted) [reply kpops/components/base_components/base_defaults_component.py]: `git revert 301284eb02dbb707fefa5ab50476b06ebee757ff`?

Comment (sujuka99) [reply kpops/components/base_components/base_defaults_component.py]: WDYT about getting rid of [COM812](https://docs.astral.sh/ruff/rules/missing-trailing-comma/), but keeping [COM818](https://docs.astral.sh/ruff/rules/trailing-comma-on-bare-tuple/) and [COM819](https://docs.astral.sh/ruff/rules/prohibited-trailing-comma/)

Comment (disrupted) [reply kpops/components/base_components/base_defaults_component.py]: good idea, that looks ideal

File: tests/component_handlers/schema_handler/test_schema_handler.py (hunk 1 above)
Comment (sujuka99) [reply tests/component_handlers/schema_handler/test_schema_handler.py:149]: :thinking: That's quite nice, it is probably the better approach imo

File: kpops/cli/main.py (hunk 3 above)
Comment (sujuka99) [reply kpops/cli/main.py:129]: Ah, the whole time I thought we had [this discussion](https://github.com/bakdata/kpops/pull/363#discussion_r1338252174) here.

**Comment by disrupted** [comment]
would it be possible to have a _problem matcher_ for Ruff where it directly annotates the errors in the PR diff? similar to [actionlint](https://github.com/rhysd/actionlint/blob/main/docs/usage.md#problem-matchers)

File: kpops/cli/main.py (hunk 3 above)
Comment (sujuka99) [reply kpops/cli/main.py]: What a nightmare... Had to cherry-pick commits one-by-one after the `revert`

Not sure what I did wrong, really tried to only revert the comma related stuff, but to no avail

Comment (sujuka99) [reply kpops/cli/main.py]: Should be good now tho

**Comment by sujuka99** [comment]
> would it be possible to have a _problem matcher_ for Ruff where it directly annotates the errors in the PR diff? similar to [actionlint](https://github.com/rhysd/actionlint/blob/main/docs/usage.md#problem-matchers)

Should be, should I try to find or write one in this PR?

**Comment by disrupted** [comment]
> > would it be possible to have a _problem matcher_ for Ruff where it directly annotates the errors in the PR diff? similar to [actionlint](https://github.com/rhysd/actionlint/blob/main/docs/usage.md?rgh-link-date=2023-09-27T10%3A31%3A17Z#problem-matchers)
> 
> Should be, should I try to find or write one in this PR?

that would be great. I know that for example Home Assistant has one if you wanna try that

File: kpops/component_handlers/helm_wrapper/helm.py [hunk 5]
```diff
@@ -31,17 +29,16 @@ def __init__(self, helm_config: HelmConfig) -> None:
...
+            raise RuntimeError(
+                f"The supported Helm version is 3.x.x. The current Helm version is {self._version.major}.{self._version.minor}.{self._version.patch}"
+            )
 
     def add_repo(
         self,
         repository_name: str,
         repository_url: str,
-        repo_auth_flags: RepoAuthFlags | None = None,
+        repo_auth_flags: RepoAuthFlags = RepoAuthFlags(),
```
Comment (disrupted) [inline kpops/component_handlers/helm_wrapper/helm.py:40]: please avoid changing any logic in this PR if possible

Comment (disrupted) [reply kpops/component_handlers/helm_wrapper/helm.py:40]: ah, you're just doing that to test the problem matcher, right?

**Comment by sujuka99** [comment]
[The problem matcher seems to work now](https://github.com/bakdata/kpops/pull/363/commits/ed8b99d49589f54f4bc2649f78eb59d570c7645f#diff-1ca7465daf9238dfe221304df636d96704ebf4935288b854a5c4c6b3ea9e3162)

File: pyproject.toml [hunk 6]
```diff
@@ -67,6 +66,99 @@ mike = "^1.1.2"
...
+  "PLW2901", # `for` loop variable `{var}` overwritten by assignment target -- Inconvenient to enforce
+  "TRY002", # Create your own exception -- Inconvenient to enforce
+  "TRY003", # Avoid specifying long messages outside the exception class -- Inconvenient to enforce
+]
+select = [
+  "F", # Pyflakes
+  "E", # pycodestyle Errors
+  "W", # pycodestyle Warnings
+  "C90", # mccabe
+  "I", # isort
```
Comment (disrupted) [inline pyproject.toml:115]: should we set
```toml
[tool.ruff.isort]
split-on-trailing-comma = false
```
to preserve the current behavior?
https://docs.astral.sh/ruff/settings/#isort-split-on-trailing-comma

**Comment by disrupted** [comment]
> [The problem matcher seems to work now](https://github.com/bakdata/kpops/pull/363/commits/ed8b99d49589f54f4bc2649f78eb59d570c7645f#diff-1ca7465daf9238dfe221304df636d96704ebf4935288b854a5c4c6b3ea9e3162)

is it possible to change the displayed name from _Test_ to _Ruff_?

File: README.md [hunk 7]
```diff
@@ -3,6 +3,7 @@
 [![Build status](https://github.com/bakdata/kpops/actions/workflows/ci.yaml/badge.svg)](https://github.com/bakdata/kpops/actions/workflows/ci.yaml)
 [![pypi](https://img.shields.io/pypi/v/kpops.svg)](https://pypi.org/project/kpops)
 [![versions](https://img.shields.io/pypi/pyversions/kpops.svg)](https://github.com/bakdata/kpops)
+[![Ruff](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/astral-sh/ruff/main/assets/badge/v2.json)](https://github.com/astral-sh/ruff)
```
Comment (disrupted) [inline README.md:6]: haha cool 😎

File: pyproject.toml (hunk 6 above)
Comment (sujuka99) [reply pyproject.toml:115]: Sure!

**Comment by sujuka99** [comment]
> > [The problem matcher seems to work now](https://github.com/bakdata/kpops/pull/363/commits/ed8b99d49589f54f4bc2649f78eb59d570c7645f#diff-1ca7465daf9238dfe221304df636d96704ebf4935288b854a5c4c6b3ea9e3162)
> 
> is it possible to change the displayed name from _Test_ to _Ruff_?

The only way I can see is to put Ruff into a separate job

**Comment by raminqaf** [comment]
I agree with @disrupted that we should gradually introduce the changes (fixes) made by Ruff. Maybe it would be good to ignore most of the (critical) rules at first and do them step by step in follow-up PRs to avoid breaking the code.

**Comment by disrupted** [comment]
> I agree with @disrupted that we should gradually introduce the changes (fixes) made by Ruff. Maybe it would be good to ignore most of the (critical) rules at first and do them step by step in follow-up PRs to avoid breaking the code.

I will look through all of the changes today and add comments if I run into particularly big changes. otherwise, of course it's best practice to bundle formatting changes in one big commit since it's easier to exclude from git blame.

**Comment by disrupted** [comment]
> > is it possible to change the displayed name from _Test_ to _Ruff_?
> 
> The only way I can see is to put Ruff into a separate job

ok, then it's fine

File: .github/ruff-matcher.json
Comment (disrupted) [inline .github/ruff-matcher.json:1]: do you think there would be an advantage of using separate matchers for warnings and errors similar to how [Home Assistant configures it](https://github.com/home-assistant/core/blob/217a895b5a0ffe67ae17f184ca577b4ed7cff955/.github/workflows/matchers/ruff.json)?

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -35,11 +35,17 @@ jobs:
...
 
-      - name: Lint (flake8)
-        run: poetry run pre-commit run flake8 --all-files --show-diff-on-failure
-
-      - name: Order of imports (isort)
-        run: poetry run pre-commit run isort --all-files --show-diff-on-failure
+      - name: Lint (ruff)
+        shell: bash
+        run: |
+          if [[ "$RUNNER_OS" == "Linux" && $(python -V) == *"3.10"* ]];
```
Comment (disrupted) [inline .github/workflows/ci.yaml:41]:
```suggestion
          if [[ "$RUNNER_OS" == "Linux" && "${{ matrix.python-version }}" == "3.10" ]];
```

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -1,10 +1,23 @@
...
-        exclude: ^tests/.*snapshots/
+      - id: ruff
+        name: ruff
+        entry: ruff check .
+        args: [ --config, pyproject.toml, --fix, --show-fixes, --exit-non-zero-on-fix ]
+        language: system
+        types_or: [python]
+        require_serial: true # run once for all files
+        pass_filenames: false
+      - id: ruff-ci
```
Comment (disrupted) [inline .pre-commit-config.yaml:12]: this one is only used for the one CI job with the problem matcher enabled, right? does it then even need to be in pre-commit?

File: kpops/component_handlers/__init__.py [hunk 10]
```diff
@@ -2,11 +2,10 @@
 
 from typing import TYPE_CHECKING
 
-from kpops.component_handlers.kafka_connect.kafka_connect_handler import (
-    KafkaConnectHandler,
-)
-
 if TYPE_CHECKING:
+    from kpops.component_handlers.kafka_connect.kafka_connect_handler import (
+        KafkaConnectHandler,
```
Comment (disrupted) [inline kpops/component_handlers/__init__.py:7]: get rid of the trailing comma

File: kpops/component_handlers/helm_wrapper/helm.py [hunk 11]
```diff
@@ -29,16 +31,17 @@ def __init__(self, helm_config: HelmConfig) -> None:
...
 
     def add_repo(
         self,
         repository_name: str,
         repository_url: str,
-        repo_auth_flags: RepoAuthFlags = RepoAuthFlags(),
+        repo_auth_flags: RepoAuthFlags | None = None,
     ) -> None:
+        if repo_auth_flags is None:
+            repo_auth_flags = RepoAuthFlags()
```
Comment (disrupted) [inline kpops/component_handlers/helm_wrapper/helm.py:44]: good! 👏🏻

File: kpops/component_handlers/helm_wrapper/helm.py [hunk 12]
```diff
@@ -206,8 +212,8 @@ def __execute(self, command: list[str]) -> str:
         log.debug(f"Executing {' '.join(command)}")
         process = subprocess.run(
             command,
-            stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE,
+            check=True,
+            capture_output=True,
```
Comment (disrupted) [inline kpops/component_handlers/helm_wrapper/helm.py:216]: did you verify that those are equal?

File: kpops/utils/dict_differ.py [hunk 13]
```diff
@@ -53,9 +57,9 @@ def from_dicts(
         d1: dict, d2: dict, ignore: set[str] | None = None
     ) -> Iterator[Diff]:
         for diff_type, keys, changes in diff(d1, d2, ignore=ignore):
-            if not isinstance(changes, list):
-                changes = [("", changes)]
-            for key, change in changes:
+            if not isinstance(changes_tmp := changes, list):
+                changes_tmp = [("", changes)]
+            for key, change in changes_tmp:
```
Comment (disrupted) [inline kpops/utils/dict_differ.py:62]: could you explain this change?

File: pyproject.toml [hunk 14]
```diff
@@ -67,6 +66,102 @@ mike = "^1.1.2"
...
+"tests/*/__init__.py" = ["F401"]
+
+[tool.ruff.isort]
+split-on-trailing-comma = false
+
+[tool.ruff.flake8-bugbear]
+extend-immutable-calls = ["typer.Argument"]
+
+[tool.ruff.flake8-type-checking]
+runtime-evaluated-base-classes = ["pydantic.BaseModel", "kpops.components.base_components.kafka_app.KafkaApp"]
```
Comment (disrupted) [inline pyproject.toml:163]: why KafkaApp?

File: tests/cli/snapshots/snap_test_schema_generation.py
Comment (disrupted) [inline tests/cli/snapshots/snap_test_schema_generation.py:1]: don't format the snapshots, they're automatically generated

File: tests/cli/test_schema_generation.py [hunk 15]
```diff
@@ -118,7 +120,6 @@ def test_gen_pipeline_schema_no_components(self):
                 ],
                 catch_exceptions=False,
             )
-            assert result.exit_code == 1
```
Comment (disrupted) [inline tests/cli/test_schema_generation.py:121]: why did you remove the assertion? probably it should just be moved outside the context manager, no?

File: tests/compiler/test_pipeline_name.py [hunk 16]
```diff
@@ -8,49 +8,51 @@
...
-    assert "random" == ENV["pipeline_name_1"]
-    assert "path" == ENV["pipeline_name_2"]
-    assert "for" == ENV["pipeline_name_3"]
-    assert "testing" == ENV["pipeline_name_4"]
+    assert ENV["pipeline_name"] == "some-random-path-for-testing"
+    assert ENV["pipeline_name_0"] == "some"
+    assert ENV["pipeline_name_1"] == "random"
+    assert ENV["pipeline_name_2"] == "path"
+    assert ENV["pipeline_name_3"] == "for"
+    assert ENV["pipeline_name_4"] == "testing"
```
Comment (disrupted) [inline tests/compiler/test_pipeline_name.py:22]: thanks, that was bothering me too 😅

File: tests/component_handlers/schema_handler/test_schema_handler.py [hunk 17]
```diff
@@ -116,45 +116,48 @@ def test_should_raise_value_error_if_schema_provider_class_not_found():
...
```
Comment (disrupted) [inline tests/component_handlers/schema_handler/test_schema_handler.py:144]: add type annotation

File: tests/pipeline/snapshots/snap_test_example.py
Comment (disrupted) [inline tests/pipeline/snapshots/snap_test_example.py:1]: don't format

File: tests/pipeline/snapshots/snap_test_pipeline.py
Comment (disrupted) [inline tests/pipeline/snapshots/snap_test_pipeline.py:1]: don't format

**Review by disrupted — DISMISSED** [review]
looks great 🤩 
(almost) no complaints :)

File: kpops/utils/dict_differ.py (hunk 13 above)
Comment (sujuka99) [reply kpops/utils/dict_differ.py:62]: [PLW2901](https://docs.astral.sh/ruff/rules/redefined-loop-name/) complained. I was not sure whether to suppress it, but ended up following the advice.

File: .github/ruff-matcher.json
Comment (sujuka99) [reply .github/ruff-matcher.json:1]: Is the matcher output supposed to be read mainly by reviewers? If I could run the same matcher locally, it would be quite nice, otherwise it may not be as useful. (to me, that is)

We would have to decide for each rule-group whether it is a warning or an error, so the regex would become rather long with our current rule set and I'm not sure how maintaining it would be like.

We could only specify the shorter list and include/exclude? Gotta remember to keep it up to date after each rule addition tho :x

File: .pre-commit-config.yaml (hunk 9 above)
Comment (sujuka99) [reply .pre-commit-config.yaml:12]: Good point, it doesn't, will fix

File: .github/ruff-matcher.json
Comment (disrupted) [reply .github/ruff-matcher.json:1]: ah, that would be quite inconvenient. Then let's keep your more inclusive/robust regex.
yeah, mostly during review

File: kpops/component_handlers/helm_wrapper/helm.py (hunk 12 above)
Comment (sujuka99) [reply kpops/component_handlers/helm_wrapper/helm.py:216]: `check` is an addition, [adds some safety](https://docs.astral.sh/ruff/rules/subprocess-run-without-check/)

`capture_output` is equivalent to the removed lines

File: pyproject.toml (hunk 14 above)
Comment (sujuka99) [reply pyproject.toml:162]: Not sure why that is the fix, but otherwise [TCH001](https://docs.astral.sh/ruff/rules/typing-only-first-party-import/) moves a couple imports into a `if TYPE_CHECKING: ...` block and breaks KPOps. `pydantic.BaseModel` should be enough to prevent that from happening I think, but maybe the `ABC` parent of `KafkaApp` throws `ruff` off.

File: tests/cli/test_schema_generation.py (hunk 15 above)
Comment (sujuka99) [reply tests/cli/test_schema_generation.py:121]: :thinking: true, I don't remember why I removed, probably didn't intend to.

Will bring it back.

Comment (sujuka99) [reply tests/cli/test_schema_generation.py:121]: I read a bit and I think it could be that we don't need that assertion. We expect an exception, so either we get it and the assertion isn't reached or we don't get the exception and the test fails anyway.

Comment (sujuka99) [reply tests/cli/test_schema_generation.py:121]: [See here](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.raises):

> When using pytest.raises as a context manager, it’s worthwhile to note that normal context manager rules apply and that the exception raised must be the final line in the scope of the context manager. Lines of code after that, within the scope of the context manager will not be executed. For example:

```python
>>> value = 15
>>> with pytest.raises(ValueError) as exc_info:
...     if value > 10:
...         raise ValueError("value must be <= 10")
...     assert exc_info.type is ValueError  # this will not execute
Instead, the following approach must be taken (note the difference in scope):
```
```python
>>> with pytest.raises(ValueError) as exc_info:
...     if value > 10:
...         raise ValueError("value must be <= 10")
...
>>> assert exc_info.type is ValueError
```

Comment (disrupted) [reply tests/cli/test_schema_generation.py:121]: makes sense

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (sujuka99) [reply .github/workflows/ci.yaml:50]: https://github.com/bakdata/kpops/pull/365

File: tests/component_handlers/schema_handler/test_schema_handler.py (hunk 1 above)
Comment (sujuka99) [reply tests/component_handlers/schema_handler/test_schema_handler.py]: Outright enabling `strict` mode produced 999 errors, 
```toml
[tool.pyright]
strict = ["*"]
```
while enabling just the following 5 rules produced 849:
```toml
[tool.pyright]
reportUnknownArgumentType = "error"
reportUnknownLambdaType = "error"
reportUnknownMemberType = "error"
reportUnknownParameterType = "error"
reportUnknownVariableType = "error"
```

[Pyright configuration docs](https://microsoft.github.io/pyright/#/configuration)

File: pyproject.toml (hunk 14 above)
Comment (disrupted) [reply pyproject.toml:162]: perhaps I am getting a slightly different error, it says to move `StreamsAppConfig` (in `StreamsApp`) and `ProducerValues` (in `StreamsApp`) (*weird naming discrepancy btw) in a type checking block. would that be fine?

Comment (sujuka99) [reply pyproject.toml:162]: It's the same error, it caused crashes when I tried, no idea why tbh

Comment (disrupted) [reply pyproject.toml:162]: what do you mean by crashes? kpops runtime errors?

Comment (sujuka99) [reply pyproject.toml:162]: yes

Comment (sujuka99) [reply pyproject.toml:162]: If you remove `KafkaApp` from the list and let `ruff` autofix the `TCH001` errors with `ruff check . --fix` `pytest` would fail

Comment (disrupted) [reply pyproject.toml:162]: I think we should open an issue for that, as you said flagging `pydantic.BaseModel` should suffice

Comment (sujuka99) [reply pyproject.toml:162]: Here is an example: `FAILED tests/components/test_streams_app.py::TestStreamsApp::test_set_topics - pydantic.errors.ConfigError: field "app" not yet prepared so type is still a ForwardRef, you might need to call StreamsApp.update_for
ward_refs().`

Comment (sujuka99) [reply pyproject.toml:162]: > I think we should open an issue for that, as you said flagging `pydantic.BaseModel` should suffice. I could imagine that the parent class resolving has a max depth or something similar which causes them not to be detected as children

Could be

Comment (sujuka99) [reply pyproject.toml:162]: The problem persists if you replace `KafkaApp` with `PipelineComponent` or `KubernetesApp`

Comment (sujuka99) [reply pyproject.toml:162]: Shouldn't be a depth issue in that case I think

Comment (sujuka99) [reply pyproject.toml:162]: Also persists with `StreamsApp` in place of `KafkaApp`

Comment (disrupted) [reply pyproject.toml:162]: yep, that's a good title

File: .github/workflows/ci.yaml [hunk 18]
```diff
@@ -59,11 +64,6 @@ jobs:
       - name: Generate pipeline definitions
         run: poetry run pre-commit run gen-docs-components --all-files --show-diff-on-failure
 
-      # TODO: enable when PEP 604 incompatibilty is in typer is resolved https://github.com/tiangolo/typer/issues/348
```
Comment (disrupted) [inline .github/workflows/ci.yaml:62]: is this enabled now in Ruff? I believe Typer still doesn't support the new type union syntax

Comment (sujuka99) [reply .github/workflows/ci.yaml:62]: It is not, but I have excluded only that particular rule from pyupgrade

Comment (sujuka99) [reply .github/workflows/ci.yaml:62]: https://github.com/bakdata/kpops/pull/363/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R96

File: pyproject.toml (hunk 14 above)
Comment (sujuka99) [reply pyproject.toml:162]: when I remove it the error no longer appears

Comment (disrupted) [reply pyproject.toml:162]: ok, then let's do it like this for now. still sounds like a bug that we should file an issue for, because `StreamsAppConfig` and `ProducerValues` weren't resolved as children of `pydantic.BaseModel`

Comment (sujuka99) [reply pyproject.toml:162]: Do you mean that I should get rid of the import?

File: tests/component_handlers/schema_handler/test_schema_handler.py (hunk 1 above)
Comment (sujuka99) [reply tests/component_handlers/schema_handler/test_schema_handler.py]: #367

**Review by disrupted — APPROVED** [review]
